igwn_auth_utils/requests.py

```diff
@@ -0,0 +1,289 @@
+            )
+        if force_noauth:
+            return
+
+        # find creds if we can
+        self._init_auth(
+            auth=auth,
+            cert=cert,
+            token=token,
+            token_audience=token_audience,
+            token_scope=token_scope,
+            url=url,
+        )
+
+        # if no auth was found, and we need it, fail here
+        if not self._has_igwn_auth and fail_if_noauth:
+            raise ValueError("no valid authorisation credentials found")
+
+    def _init_auth(
+        self,
+        auth=None,
+        cert=None,
+        token=None,
+        token_audience=None,
+        token_scope=None,
+        url=None,
+    ):
+        # -- bearer token
+        tokenhdr = self._find_token_header(
+            token,
+            token_audience,
+            token_scope,
+        )
+        if tokenhdr:
+            self.headers["Authorization"] = tokenhdr
+            return
+
+        # -- cert auth
+        self.cert = self._find_x509_credentials(cert)
+        if self.cert:
+            return
+
+        # -- basic auth (netrc)
+        self.auth = self._find_username_password(url)
+
+    @property
+    def _has_igwn_auth(self):
+        """`True` if this `Session` is configured for IGWN authorization,
+        otherwise `False`.
+
+        :type: `bool`
+        """
+        return any((
+            self.auth,
+            "Authorization" in self.headers,
+            self.cert,
+        ))
+
+    @staticmethod
+    def _find_x509_credentials(cert):
+        """Find an X.509 certificate for authorization
+        """
+        if cert is False:  # disable
+            return None
+        if cert not in {True, None}:  # user gave us a cert
+            return cert
+        try:
+            return find_x509_credentials()
+        except RuntimeError:
+            # user asked for a cert, so raise errors finding one
+            if cert is True:
+                raise
+            # otherwise ignore
+            return
+
+    @staticmethod
+    def _find_token_header(token, audience, scope):
+        """Find a bearer token for authorization
+        """
+        if token is False:  # disable
+            return
+        if token in {True, None}:  # user didn't pass a token
+            try:
+                token = find_scitoken(audience, scope)
+            except RuntimeError:
+                # user asked to use scitokens, so raise errors finding one
+                if token is True:
+                    raise
+                # otherwise ignore
+                return
+        # return the authorization header for the token
+        return scitoken_authorization_header(token)
+
+    @staticmethod
+    def _find_username_password(url):
+        """Use `netrc.netrc` to find the username/password for basic auth
+        """
+        host = urlparse(url).hostname
+
+        try:
+            result = netrc(getenv("NETRC")).authenticators(host)
+        except IOError:
+            return
```

Comment:
Not `raise`?

Reply:
No, in line with the other `_find_xxx` methods it doesn't ever fail if credentials aren't found.